File: backend/services/pdf_service.py
import io
import os
import random
from fpdf import FPDF

FONTS_DIR = os.path.join(os.path.dirname(__file__), '..', 'fonts')
FONT_REGULAR = os.path.join(FONTS_DIR, 'NotoSans-Regular.ttf')
FONT_BOLD    = os.path.join(FONTS_DIR, 'NotoSans-Bold.ttf')

LABELS = ['A', 'B', 'C', 'D']


# ── Shuffle ───────────────────────────────────────────────────────────────────

def _shuffle(lst):
    result = lst[:]
    random.shuffle(result)
    return result


def generate_variants(questions: list, count: int) -> list:
    variants = []
    for i in range(count):
        code = str(i + 1).zfill(3)
        shuffled_qs = _shuffle(questions)
        variant_questions = []
        for q in shuffled_qs:
            shuffled_answers = _shuffle(q['answers'])
            correct_idx = next(j for j, a in enumerate(shuffled_answers) if a['is_correct'])
            variant_questions.append({
                'question_id':  q['question_id'],
                'content':      q['content'],
                'answers':      [
                    {'answer_id': a['answer_id'], 'content': a['content'], 'label': LABELS[j]}
                    for j, a in enumerate(shuffled_answers)
                ],
                'correct_label': LABELS[correct_idx],
            })
        variants.append({'code': code, 'questions': variant_questions})
    return variants


# ── PDF builder ───────────────────────────────────────────────────────────────

class ExamPDF(FPDF):
    def __init__(self):
        super().__init__(orientation='P', unit='mm', format='A4')
        self.add_font('NotoSans',      style='',  fname=FONT_REGULAR)
        self.add_font('NotoSans',      style='B', fname=FONT_BOLD)
        self.set_auto_page_break(auto=True, margin=18)
        self.set_margins(left=20, top=20, right=20)

    def _divider(self, thick=False):
        self.set_draw_color(0, 0, 0)
        self.set_line_width(0.4 if thick else 0.2)
        self.line(self.get_x(), self.get_y(), self.w - 20, self.get_y())
        self.ln(3)

    def _info_line(self, label: str, width: float):
        self.set_font('NotoSans', 'B', 10)
        self.cell(self.get_string_width(label) + 1, 6, label)
        x_start = self.get_x()
        self.set_draw_color(100, 100, 100)
        self.set_line_width(0.2)
        self.line(x_start, self.get_y() + 5, x_start + width, self.get_y() + 5)
        self.cell(width + 2, 6, '')

    def add_exam_page(self, exam_name: str, duration: int, variant: dict):
        self.add_page()

        # ── Header ────────────────────────────────────────────────────────────
        self.set_font('NotoSans', '', 9)
        self.set_text_color(100, 100, 100)
        self.cell(0, 5, 'PHIẾU KIỂM TRA TOÁN LỚP 3', align='L', new_x='LMARGIN', new_y='NEXT')
        self.set_text_color(0, 0, 0)

        # Exam name (centered) + code box (right)
        self.set_font('NotoSans', 'B', 13)
        name_w = self.w - 40 - 40  # leave space for code box on the right
        self.cell(name_w, 7, exam_name, align='C')
        # Mã đề box
        code_text = f'Mã đề: {variant["code"]}'
        code_w    = self.get_string_width(code_text) + 8
        self.set_font('NotoSans', 'B', 11)
        self.set_draw_color(0, 0, 0)
        self.set_line_width(0.4)
        self.cell(code_w, 7, code_text, border=1, align='C', new_x='LMARGIN', new_y='NEXT')

        self.set_font('NotoSans', '', 10)
        self.set_text_color(60, 60, 60)
        self.cell(0, 5, f'Thời gian làm bài: {duration} phút', align='C', new_x='LMARGIN', new_y='NEXT')
        self.set_text_color(0, 0, 0)
        self.ln(2)

        self._divider(thick=True)

        # ── Student info ──────────────────────────────────────────────────────
        self._info_line('Họ và tên:', 80)
        self._info_line('Lớp:', 30)
        self.ln(8)
        self._info_line('Ngày kiểm tra:', 50)
        self._info_line('Điểm:', 25)
        self.ln(10)

        self._divider(thick=True)
        self.ln(2)

        # ── Questions ─────────────────────────────────────────────────────────
        for idx, q in enumerate(variant['questions'], start=1):
            # Question text
            self.set_font('NotoSans', 'B', 10)
            q_prefix = f'Câu {idx}: '
            prefix_w = self.get_string_width(q_prefix)
            self.cell(prefix_w, 6, q_prefix)
            self.set_font('NotoSans', '', 10)
            self.multi_cell(0, 6, q['content'], new_x='LMARGIN', new_y='NEXT')

            # Answers in 2-column grid
            col_w = (self.w - 40) / 2
            for i, ans in enumerate(q['answers']):
                if i % 2 == 0 and i != 0:
                    self.ln(0)
                x_offset = 5 + (col_w * (i % 2))
                self.set_x(self.l_margin + x_offset)
                self.set_font('NotoSans', 'B', 10)
                label_text = f'{ans["label"]}. '
                self.cell(self.get_string_width(label_text), 6, label_text)
                self.set_font('NotoSans', '', 10)
                self.cell(col_w - self.get_string_width(label_text) - 2, 6, ans['content'],
                          new_x='RIGHT' if i % 2 == 0 else 'LMARGIN',
                          new_y='LAST' if i % 2 == 0 else 'NEXT')

            if len(q['answers']) % 2 != 0:
                self.ln(6)
            self.ln(3)

    def add_answer_key_page(self, exam_name: str, variants: list):
        self.add_page()

        self.set_font('NotoSans', 'B', 14)
        self.cell(0, 8, 'BẢNG ĐÁP ÁN', align='C', new_x='LMARGIN', new_y='NEXT')
        self.set_font('NotoSans', '', 10)
        self.set_text_color(100, 100, 100)
        self.cell(0, 5, exam_name, align='C', new_x='LMARGIN', new_y='NEXT')
        self.set_text_color(0, 0, 0)
        self.ln(4)

        self._divider(thick=False)

        # Table header
        self.set_fill_color(243, 244, 246)
        self.set_font('NotoSans', 'B', 10)
        self.cell(22, 7, 'Mã đề', border='B', fill=True)
        self.cell(0,  7, 'Đáp án', border='B', fill=True, new_x='LMARGIN', new_y='NEXT')

        # Rows
        self.set_font('NotoSans', '', 10)
        for variant in variants:
            answers_str = '   '.join(
                f"{i + 1}.{q['correct_label']}"
                for i, q in enumerate(variant['questions'])
            )
            self.set_draw_color(220, 220, 220)
            self.set_line_width(0.2)
            self.cell(22, 7, variant['code'], border='B')
            self.multi_cell(0, 7, answers_str, border='B', new_x='LMARGIN', new_y='NEXT')


def build_exam_pdf(exam_name: str, duration: int, questions: list, variants_count: int) -> bytes:
    variants = generate_variants(questions, variants_count)

    pdf = ExamPDF()
    for variant in variants:
        pdf.add_exam_page(exam_name, duration, variant)
    pdf.add_answer_key_page(exam_name, variants)

    buf = io.BytesIO()
    pdf.output(buf)
    return buf.getvalue()

File: backend/src/services/pdf.service.js
import PDFDocument from 'pdfkit';
import { fileURLToPath } from 'url';
import path from 'path';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const FONTS_DIR = path.join(__dirname, '..', '..', 'fonts');
const FONT_REGULAR = path.join(FONTS_DIR, 'NotoSans-Regular.ttf');
const FONT_BOLD = path.join(FONTS_DIR, 'NotoSans-Bold.ttf');

const LABELS = ['A', 'B', 'C', 'D'];

function shuffle(arr) {
  const result = [...arr];
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [result[i], result[j]] = [result[j], result[i]];
  }
  return result;
}

function generateVariants(questions, count) {
  const variants = [];
  for (let i = 0; i < count; i++) {
    const code = String(i + 1).padStart(3, '0');
    const shuffledQs = shuffle(questions);
    const variantQuestions = shuffledQs.map(q => {
      const shuffledAnswers = shuffle(q.answers);
      const correctIdx = shuffledAnswers.findIndex(a => a.is_correct);
      return {
        question_id: q.question_id,
        content: q.content,
        answers: shuffledAnswers.map((a, j) => ({ ...a, label: LABELS[j] })),
        correct_label: LABELS[correctIdx],
      };
    });
    variants.push({ code, questions: variantQuestions });
  }
  return variants;
}

export const buildExamPdf = (examName, duration, questions, variantsCount) => {
  return new Promise((resolve, reject) => {
    const variants = generateVariants(questions, variantsCount);
    const chunks = [];

    const doc = new PDFDocument({
      size: 'A4',
      margin: 50,
      autoFirstPage: false,
    });

    doc.registerFont('NotoSans', FONT_REGULAR);
    doc.registerFont('NotoSans-Bold', FONT_BOLD);

    doc.on('data', chunk => chunks.push(chunk));
    doc.on('end', () => resolve(Buffer.concat(chunks)));
    doc.on('error', reject);

    const pageW = 595.28;
    const margin = 50;
    const contentW = pageW - margin * 2;

    for (const variant of variants) {
      doc.addPage();
      let y = margin;

      doc.font('NotoSans').fontSize(8).fillColor('#888').text('PHIẾU KIỂM TRA TOÁN LỚP 3', margin, y);
      y += 16;

      doc.font('NotoSans-Bold').fontSize(13).fillColor('#000');
      const codeText = `Mã đề: ${variant.code}`;
      const codeW = doc.widthOfString(codeText) + 16;
      const nameW = contentW - codeW - 8;
      doc.text(examName, margin, y, { width: nameW, align: 'center' });
      doc.rect(margin + nameW + 8, y - 1, codeW, 16).stroke();
      doc.font('NotoSans-Bold').fontSize(11).text(codeText, margin + nameW + 12, y + 1);
      y += 22;

      doc.font('NotoSans').fontSize(10).fillColor('#444')
        .text(`Thời gian làm bài: ${duration} phút`, margin, y, { width: contentW, align: 'center' });
      y += 16;

      doc.moveTo(margin, y).lineTo(margin + contentW, y).lineWidth(1).stroke('#000');
      y += 8;

      doc.font('NotoSans-Bold').fontSize(10).fillColor('#000').text('Họ và tên:', margin, y);
      const nameLineX = margin + doc.widthOfString('Họ và tên:') + 4;
      doc.moveTo(nameLineX, y + 12).lineTo(nameLineX + 180, y + 12).lineWidth(0.5).stroke('#888');
      doc.font('NotoSans-Bold').text('Lớp:', nameLineX + 190, y);
      const classLineX = nameLineX + 190 + doc.widthOfString('Lớp:') + 4;
      doc.moveTo(classLineX, y + 12).lineTo(classLineX + 60, y + 12).lineWidth(0.5).stroke('#888');
      y += 20;

      doc.font('NotoSans-Bold').fontSize(10).fillColor('#000').text('Ngày kiểm tra:', margin, y);
      const dateLineX = margin + doc.widthOfString('Ngày kiểm tra:') + 4;
      doc.moveTo(dateLineX, y + 12).lineTo(dateLineX + 120, y + 12).lineWidth(0.5).stroke('#888');
      doc.font('NotoSans-Bold').text('Điểm:', dateLineX + 130, y);
      const scoreLineX = dateLineX + 130 + doc.widthOfString('Điểm:') + 4;
      doc.moveTo(scoreLineX, y + 12).lineTo(scoreLineX + 60, y + 12).lineWidth(0.5).stroke('#888');
      y += 24;

      doc.moveTo(margin, y).lineTo(margin + contentW, y).lineWidth(1).stroke('#000');
      y += 10;

      const colW = contentW / 2;
      for (let qi = 0; qi < variant.questions.length; qi++) {
        const q = variant.questions[qi];

        if (y > 750) { doc.addPage(); y = margin; }

        doc.font('NotoSans-Bold').fontSize(10).fillColor('#000');
        const prefix = `Câu ${qi + 1}: `;
        const prefixW = doc.widthOfString(prefix);
        doc.text(prefix, margin, y, { continued: true });
        doc.font('NotoSans').text(q.content, { width: contentW - prefixW });
        y = doc.y + 4;

        for (let ai = 0; ai < q.answers.length; ai++) {
          if (y > 760) { doc.addPage(); y = margin; }
          const ans = q.answers[ai];
          const xOffset = margin + (ai % 2 === 0 ? 8 : colW + 8);
          doc.font('NotoSans-Bold').fontSize(10).text(`${ans.label}. `, xOffset, y, { continued: true });
          doc.font('NotoSans').text(ans.content, { width: colW - 16 });
          if (ai % 2 === 1 || ai === q.answers.length - 1) {
            y = doc.y + 2;
          }
        }
        if (q.answers.length % 2 !== 0) y = doc.y + 2;
        y += 6;
      }
    }

    doc.addPage();
    let y = margin;
    doc.font('NotoSans-Bold').fontSize(14).fillColor('#000')
      .text('BẢNG ĐÁP ÁN', margin, y, { width: contentW, align: 'center' });
    y += 20;
    doc.font('NotoSans').fontSize(10).fillColor('#888')
      .text(examName, margin, y, { width: contentW, align: 'center' });
    y += 20;

    doc.moveTo(margin, y).lineTo(margin + contentW, y).lineWidth(0.5).stroke('#ccc');
    y += 8;

    const col1W = 60;
    doc.font('NotoSans-Bold').fontSize(10).fillColor('#000')
      .text('Mã đề', margin, y, { width: col1W })
      .text('Đáp án', margin + col1W, y);
    y += 18;

    for (const variant of variants) {
      if (y > 760) { doc.addPage(); y = margin; }
      const answersStr = variant.questions.map((q, i) => `${i + 1}.${q.correct_label}`).join('   ');
      doc.font('NotoSans').fontSize(10)
        .text(variant.code, margin, y, { width: col1W })
        .text(answersStr, margin + col1W, y, { width: contentW - col1W });
      y = doc.y + 4;
      doc.moveTo(margin, y).lineTo(margin + contentW, y).lineWidth(0.3).stroke('#ddd');
      y += 4;
    }

    doc.end();
  });
};
